In [1]:
# import necessary libraries

import pandas as pd
import seaborn as sns

## Preliminary Data Inspection

This class is designed to streamline the initial exploration and quality assessment of a daraset. Providing methods like loading the datatset, inspecting the structure, previewing, identifying missing values and detecting duplicates.



In [2]:
# Create a class that handles Prelimary Data Inspection

class PreliminaryDataInspection():
    '''
    This class performs preliminary data inspection.
    '''
    def __init__(self, file_path, sep=","):
        '''
        Initializes the class by loading the DataFrame.

        Args:
            file_path (str): Path to the data file.
            sep (str): Separator used in the file. Default is ",".
        '''
        self.df = self.load_data(file_path, sep)

    def load_data(self, file_path, sep=","):
        '''
        Loads a data file and returns a DataFrame.

        Args:
            file_path (str): Path to the data file.
            sep (str): Separator used in the file. Default is ",".

        Returns:
            pd.DataFrame or None: The loaded DataFrame or None if file not found.
        '''
        try:
            df = pd.read_csv(file_path, sep=sep)
            print(f"Data successfully loaded from {file_path}.")
            return df
        except FileNotFoundError:
            print(f"File not found: {file_path}")
            return None
        except Exception as e:
            print(f"An error occurred: {e}")
            return None

    def check_info(self):
        '''
        Displays information about the DataFrame.
        '''
        if self.df is not None:
            print("DataFrame Info:")
            print(self.df.info())
        else:
            print("No DataFrame to inspect.")

    def check_head(self, n=5):
        '''
        Displays the first few rows of the DataFrame.

        Args:
            n (int): Number of rows to display. Default is 5.
        '''
        if self.df is not None:
            print(f"First {n} rows of the DataFrame:")
            print(self.df.head(n))
        else:
            print("No DataFrame to inspect.")

    def check_missing_values(self):
        '''
        Checks for missing values in the DataFrame.
        '''
        if self.df is not None:
            print("Missing Values in DataFrame:")
            missing_values = self.df.isnull().sum()
            print(missing_values[missing_values > 0])
        else:
            print("No DataFrame to inspect.")

    def check_duplicates(self):
        '''
        Checks for duplicate rows in the DataFrame.
        '''
        if self.df is not None:
            duplicates = self.df.duplicated().sum()
            print(f"Number of duplicate rows: {duplicates}")
        else:
            print("No DataFrame to inspect.")


### Alarm Response (`Alarmresponses.txt`)

In [4]:
alarm_response = PreliminaryDataInspection('data/Alarmresponses.txt', sep='\t')

alarm_response.check_info()
alarm_response.check_head()
alarm_response.check_missing_values()
alarm_response.check_duplicates()

Data successfully loaded from data/Alarmresponses.txt.
DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4124815 entries, 0 to 4124814
Data columns (total 8 columns):
 #   Column        Dtype 
---  ------        ----- 
 0   alarmid       int64 
 1   userid        int64 
 2   DeviceType    int64 
 3   devices       object
 4   Response      int64 
 5   ResponseDate  object
 6   latitude      object
 7   longitude     object
dtypes: int64(4), object(4)
memory usage: 251.8+ MB
None
First 5 rows of the DataFrame:
   alarmid               userid  DeviceType devices  Response ResponseDate  \
0    78100  2210044967363094894           4     Sms         0          NaN   
1    79629  2210044967363094894           4     Sms         0          NaN   
2    82395  2210044967363094894           4     Sms         0          NaN   
3    89313  2210044967363094894           4     Sms         0          NaN   
4    43918   738033312678202846           4     Sms         0          NaN   



### Data Summary:

Alarm reponses contains 4,124,815 entries with 8 columns (4214815 x 8 matrix), the columns are:

- alarmid: Unique identifier for each alarm event (integer).
- userid: Unique identifier for the user who triggered the alarm (integer).
- DeviceType: Type of device used to trigger the alarm (integer).
- devices: Description of the device used (object). 193 entries are missing values.
- Response: Indicates if a volunteer responded to the alarm (0 - no response, 1 - responded) (integer).
- ResponseDate: Date and time the volunteer responded (object). 3,603,938 entries are missing values.
- latitude: Latitude coordinate of the alarm location (object).
- longitude: Longitude coordinate of the alarm location (object).

**Data Insights:**

- A large portion of the data (over 87%) has missing values in the RespinseDate column. As at now, no conclusion could be made regarding the missing date values. Depending on its correlation with the target variable, I might result to dropping the column if it is negatively correlated with the Response column.

- There are missing values in the devices column, indicating some descriptions of the triggering devices are unavailable.

- The data seems suitable for analyzing volunteer response patterns, effectiveness, and identifying areas for improvement.


### Alarms (`Alarms.txt`)

In [5]:
alarms = PreliminaryDataInspection('data/Alarms.txt', sep='\t')

alarms.check_info()
alarms.check_head()
alarms.check_missing_values()
alarms.check_duplicates()

Data successfully loaded from data/Alarms.txt.
DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60329 entries, 0 to 60328
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   alarmid         60329 non-null  int64 
 1   alarmlocaltime  60329 non-null  object
 2   Latitude        60329 non-null  object
 3   Longitude       60329 non-null  object
 4   AlarmStatus     60329 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 2.3+ MB
None
First 5 rows of the DataFrame:
   alarmid           alarmlocaltime   Latitude Longitude  AlarmStatus
0    29977  2020-01-01 00:23:14.200   52,26168  5,603156            1
1    29979  2020-01-01 00:25:36.730  51,900643  4,484577            1
2    29980  2020-01-01 00:34:54.213   52,63205  4,726312            1
3    29982  2020-01-01 00:58:24.897  51,263437  4,045439            1
4    29983  2020-01-01 00:59:01.363  51,334741  6,129295            1
Missing Values 

### Data Summary:

The data is stored in a pandas DataFrame containing 60,329 entries. There are 5 columns in the DataFrame:
- alarmid: Unique identifier for each alarm event (integer).
- alarmlocaltime: Local time of the alarm event (object).
- Latitude: Latitude coordinate of the alarm location (object).
- Longitude: Longitude coordinate of the alarm location (object).
- AlarmStatus: Status of the alarm (integer, likely indicating active/inactive).

**Data Insights:**

- The data has no missing values.
- The alarmlocaltime column contains timestamp information, which can be useful for temporal analysis.
- The Latitude and Longitude columns provide geographic information, enabling spatial analysis.
- The AlarmStatus column can help identify active alarms and potentially correlate with response times or other factors.

### Volunteer availability



In [6]:
volunteer_availability = PreliminaryDataInspection('data/VolunteerAvailability.txt', sep='\t')

volunteer_availability.check_info()
volunteer_availability.check_head()
volunteer_availability.check_missing_values()
volunteer_availability.check_duplicates()

Data successfully loaded from data/VolunteerAvailability.txt.
DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2879387 entries, 0 to 2879386
Data columns (total 8 columns):
 #   Column                Dtype 
---  ------                ----- 
 0   userid                int64 
 1   FromHour              int64 
 2   FromMin               int64 
 3   ToHour                int64 
 4   ToMin                 int64 
 5   AvailableWeekDay      int64 
 6   AvailabilityLocation  int64 
 7   CreatedDate           object
dtypes: int64(7), object(1)
memory usage: 175.7+ MB
None
First 5 rows of the DataFrame:
                userid  FromHour  FromMin  ToHour  ToMin  AvailableWeekDay  \
0  2210044967363094894         0        0      23     59                 0   
1  2210044967363094894         0        0      23     59                 1   
2  2210044967363094894         0        0      23     59                 2   
3  2210044967363094894         0        0      23     59              

In [7]:
volunteers = PreliminaryDataInspection('data/Volunteers.txt', sep='\t')

volunteers.check_info()
volunteers.check_head()
volunteers.check_missing_values()
volunteers.check_duplicates()

Data successfully loaded from data/Volunteers.txt.
DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 246088 entries, 0 to 246087
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   userid         246088 non-null  int64 
 1   DateOfBirth    246088 non-null  object
 2   Latitude       246088 non-null  object
 3   Longitude      246088 non-null  object
 4   WorkLatitude   246088 non-null  object
 5   WorkLongitude  246088 non-null  object
dtypes: int64(1), object(5)
memory usage: 11.3+ MB
None
First 5 rows of the DataFrame:
                userid              DateOfBirth          Latitude  \
0 -4402443911528580314  1973-10-16 00:00:00.000         53,239679   
1  4147689039375744256  1989-07-11 00:00:00.000        52,3007863   
2 -7843085838038396722  1997-07-01 00:00:00.000  52,3327560424805   
3 -1646437989330162528  1976-01-30 00:00:00.000  52,6543579101563   
4  5678998634590506361  1962-05-14 00:0

## Create a Unified Dataset

In [8]:
merged_data = pd.merge(alarm_response.df, alarms.df, on='alarmid', how='left')
merged_data = pd.merge(merged_data, volunteers.df, on='userid', how='left')
merged_data = pd.merge(merged_data, volunteer_availability.df, on='userid', how='left')
merged_data.head()

alarmid               userid  DeviceType devices  Response ResponseDate  \
0    78100  2210044967363094894           4     Sms         0          NaN   
1    78100  2210044967363094894           4     Sms         0          NaN   
2    78100  2210044967363094894           4     Sms         0          NaN   
3    78100  2210044967363094894           4     Sms         0          NaN   
4    78100  2210044967363094894           4     Sms         0          NaN   

           latitude         longitude           alarmlocaltime Latitude_x  \
0  51,9987411499023  4,77543687820435  2023-09-08 15:19:27.160  51,999415   
1  51,9987411499023  4,77543687820435  2023-09-08 15:19:27.160  51,999415   
2  51,9987411499023  4,77543687820435  2023-09-08 15:19:27.160  51,999415   
3  51,9987411499023  4,77543687820435  2023-09-08 15:19:27.160  51,999415   
4  51,9987411499023  4,77543687820435  2023-09-08 15:19:27.160  51,999415   

   ...       Longitude_y  WorkLatitude WorkLongitude FromHour FromMin ToHour  \
0  ...  4,77543687820435             0             0      0.0     0.0   23.0   
1  ...  4,77543687820435             0             0      0.0     0.0   23.0   
2  ...  4,77543687820435             0             0      0.0     0.0   23.0   
3  ...  4,77543687820435             0             0      0.0     0.0   23.0   
4  ...  4,77543687820435             0             0      0.0     0.0   23.0   

  ToMin  AvailableWeekDay  AvailabilityLocation              CreatedDate  
0  59.0               0.0                   0.0  2023-05-25 19:59:53.413  
1  59.0               1.0                   0.0  2023-05-25 19:59:53.413  
2  59.0               2.0                   0.0  2023-05-25 19:59:53.413  
3  59.0               3.0                   0.0  2023-05-25 19:59:53.413  
4  59.0               4.0                   0.0  2023-05-25 19:59:53.413  

[5 rows x 24 columns]

## DataPreprocessor Class

In [ ]:
class DataPreprocessor:
    """
    A class for preprocessing data, including handling missing values, converting columns
    to numeric, and removing duplicates.
    """
    def __init__(self, df):
        """
        Initialize the DataPreprocessor with a DataFrame.
        """
        self.df = df

    def handle_missing_values(self, unique_identifiers):
        """
        Handle missing values in the DataFrame.

        Args:
            unique_identifiers (list): List of unique identifier column names, including the target column.
        """
        # Drop rows where unique identifiers have missing values
        for col in unique_identifiers:
            if col in self.df.columns:
                initial_count = self.df.shape[0]
                self.df = self.df.dropna(subset=[col])
                print(f"Dropped {initial_count - self.df.shape[0]} rows with missing values in {col}")

        # Impute missing values for other columns
        for col in self.df.columns:
            if col not in unique_identifiers:
                if self.df[col].dtype in ['float64', 'int64']:
                    # Use mean for numerical columns
                    mean_value = self.df[col].mean()
                    self.df[col].fillna(mean_value, inplace=True)
                    print(f"Replaced missing values in numerical column {col} with mean: {mean_value:.2f}")
                elif self.df[col].dtype == 'object':
                    # Use mode for categorical columns
                    mode_value = self.df[col].mode()[0]
                    self.df[col].fillna(mode_value, inplace=True)
                    print(f"Replaced missing values in categorical column {col} with mode: {mode_value}")

        print("Missing values handled.")

    def convert_columns_to_numeric(self, columns):
        """
        Converts specified columns to numeric data type, handling non-convertible values.

        Args:
            columns (list): List of column names to convert.
        """
        if len(columns) == 0:
            print("No columns specified for conversion.")
            return
        for col in columns:
            if col in self.df.columns:
                # Check if the column type is object (string) before applying str.replace
                if self.df[col].dtype == 'object':
                    self.df[col] = self.df[col].str.replace(",", ".")  # Replace commas with dots for conversion
                self.df[col] = pd.to_numeric(self.df[col], errors='coerce')
                print(f"Converted column '{col}' to numeric.")
            else:
                print(f"Column '{col}' not found in DataFrame.")

    def remove_duplicates(self):
        """
        Removes duplicate rows from the DataFrame.
        """
        initial_count = self.df.shape[0]
        self.df = self.df.drop_duplicates()
        print(f"Removed {initial_count - self.df.shape[0]} duplicate rows.")

    def get_cleaned_data(self):
        """
        Return the cleaned DataFrame.

        Returns:
            pd.DataFrame: The cleaned DataFrame.
        """
        return self.df

In [ ]:
# Initialize the preprocessor with the DataFrame
preprocessor = DataPreprocessor(alarm_response.df)
unique_identifiers = ['alarmid', 'userid', 'Response', 'latitude', 'longitude']

# Convert specified columns to numeric
preprocessor.convert_columns_to_numeric(columns=['latitude', 'longitude'])

# Remove duplicates
preprocessor.remove_duplicates()

# Handle missing values
preprocessor.handle_missing_values(unique_identifiers=unique_identifiers)

# Get the cleaned DataFrame
cleaned_df = preprocessor.get_cleaned_data()
cleaned_df.head()

Converted column 'latitude' to numeric.
Converted column 'longitude' to numeric.
Removed 0 duplicate rows.
Dropped 0 rows with missing values in alarmid
Dropped 0 rows with missing values in userid
Dropped 0 rows with missing values in Response
Dropped 0 rows with missing values in latitude
Dropped 0 rows with missing values in longitude
Replaced missing values in numerical column DeviceType with mean: 2.47


<ipython-input-23-81bbd9683c64>:32: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  self.df[col].fillna(mean_value, inplace=True)
<ipython-input-23-81bbd9683c64>:37: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using

Replaced missing values in categorical column devices with mode: Sms
Missing values handled.


alarmid               userid  DeviceType devices  Response   latitude  \
0    78100  2210044967363094894           4     Sms         0  51.998741   
1    79629  2210044967363094894           4     Sms         0  51.998741   
2    82395  2210044967363094894           4     Sms         0  51.998741   
3    89313  2210044967363094894           4     Sms         0  51.998741   
4    43918   738033312678202846           4     Sms         0  53.211500   

   longitude  
0   4.775437  
1   4.775437  
2   4.775437  
3   4.775437  
4   5.769591